### PR TITLE
feat: support top-p top-k in grpo

### PR DESCRIPTION
**Pull request by yuki-97**
Thanks a lot for @zhandaz for the original implementation https://github.com/NVIDIA-NeMo/RL/pull/1578 and re-implement https://github.com/NVIDIA-NeMo/RL/pull/1938. Really awesome work!
This PR re-implement this based on the refactored loss structure.

# What does this PR do ?
For the logprobs:
1. `generation_logprobs`, `curr_logprobs` and `prev_logprobs` are computed **with** filtering.
2. `reference_policy_logprobs` is computed **without** filtering.

This PR:
1. Apply top-p top-k and apply mask -inf on `prev_logprobs` at `get_logprobs`, since we need to call `compute_and_apply_seq_logprob_error_masking` before train, which needs the updated `prev_logprobs` to compare with `generation_logprobs`.
2. Apply top-p top-k and apply mask -inf on `curr_logprobs` at `prepare_loss_input`, since we handle distributed logits to logprobs here.
4. Save and restore sampling params at `use_reference_model` to make `reference_policy_logprobs` computed without filtering.
5. For TP, we convert from vocab-parallel to batch-sequence-parallel layout via all-to-all communication, apply filtering on the full vocabulary, then convert back. This avoids materializing the full vocabulary on any single rank.

More details can be seen at https://github.com/NVIDIA-NeMo/RL/pull/1578 and https://github.com/NVIDIA-NeMo/RL/pull/1938.

# Test Result
| | convergence | time cost |
|-|-|-|
| dtensorv1 | image | image |
| dtensorv2 | image | image |
| mcore | image | image |

## Summary by CodeRabbit

* **New Features**
  * Added configurable sampling parameters (temperature, top-k, top-p) for training with new YAML configuration examples.
  * Introduced memory-efficient top-k and top-p filtering for logit sampling.

* **Bug Fixes**
  * Improved token-level masking for logprob calculations in training pipelines.

* **Tests**
  * Added comprehensive test coverage for sampling parameter validation and distributed logprob computations.
  * Expanded nightly test suite with new sampling configuration scenarios.

### CODE REVIEW

File: nemo_rl/algorithms/loss/loss_functions.py [hunk 1]
```diff
@@ -273,27 +276,39 @@ def __call__(
...
+            # - For KL, we need curr_logprobs **without** filtering to be consistent with ref logprobs
+            # - For importance weights, we also use unfiltered curr_logprobs_for_kl since we're
+            #   reweighting samples from π_gen_filtered to π_curr_unfiltered
+
+            # On-policy KL approximation
             if self.use_on_policy_kl_approximation:
                 # See: docs/guides/grpo.md#on-policy-kl-approximation
                 kl_importance_weights = torch.exp(
-                    curr_logprobs - generation_logprobs
+                    curr_logprobs_unfiltered - generation_logprobs
```
Comment (terrykong) [inline nemo_rl/algorithms/loss/loss_functions.py:290]: isn't `generation_logprobs` filtered if sampling is on, so according to the comment, shouldn't this be `curr_logprobs` so that it's the ratio of filtered-to-filters vs. unfiltered-to-filters?

File: nemo_rl/models/megatron/setup.py [hunk 2]
```diff
@@ -194,12 +195,20 @@ def validate_and_set_config(
...
     if "generation" in config and config["generation"] is not None:
-        is_generation_colocated = config["generation"]["colocated"]["enabled"]
+        generation_cfg = config["generation"]
+        # set generation colocated
+        is_generation_colocated = generation_cfg["colocated"]["enabled"]
+        # set sampling params
+        sampling_params = TrainingSamplingParams(
+            top_k=generation_cfg.get("top_k", None),
+            top_p=generation_cfg.get("top_p", 1.0),
+            temperature=generation_cfg.get("temperature", 1.0),
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py]: shall we just count on those being in the cfg?

File: nemo_rl/models/policy/workers/dtensor_policy_worker.py [hunk 3]
```diff
@@ -169,8 +175,17 @@ def __init__(
...
         if "generation" in config and config["generation"] is not None:
-            self.is_generation_colocated = config["generation"]["colocated"]["enabled"]
+            generation_cfg = config["generation"]
+            # set generation colocated
+            self.is_generation_colocated = generation_cfg["colocated"]["enabled"]
+            # set sampling params
+            self.sampling_params = TrainingSamplingParams(
+                top_k=generation_cfg.get("top_k", None),
+                top_p=generation_cfg.get("top_p", 1.0),
+                temperature=generation_cfg.get("temperature", 1.0),
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker.py]: same comment about the defaults

File: nemo_rl/models/policy/workers/dtensor_policy_worker.py [hunk 4]
```diff
@@ -1615,8 +1656,10 @@ def get_topk_logits(
...
 
-        On entry: Moves model to CPU, moves reference_model to CUDA. Swaps the references
-        On exit: Restores original references and re-flips cuda/cpu
+        On entry: Moves model to CPU, moves reference_model to CUDA. Swaps the references.
+                  Also disables top-k/top-p filtering since the reference policy's distribution
+                  is different from the current policy, making filtered logprobs incompatible.
+        On exit: Restores original references and re-flips cuda/cpu, restores sampling_params.
         """
         with torch.no_grad():
             try:
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker.py]: actually is this try:finally needed? if it crashes inside, is there a point in doing the finally and then crashing?

File: nemo_rl/algorithms/utils.py [hunk 5]
```diff
@@ -182,6 +182,38 @@ def masked_mean(
...
+
+    Handling sampling mask mismatch:
+    vLLM samples token X from top-k/p filtered distribution -> generation_logprobs[X] is always finite (e.g., -5.41)
+    during training: policy computes logprobs with same top-k/p settings, but the distribution can be slightly different
+    token X may fall outside the training policy's top-k/p set -> curr_logprobs[X] = -inf, prev_logprobs[X] = -inf
+    Detect positions with -inf in any logprobs (generation_logprobs is always finite for valid tokens)
+
+    Args:
+        logprobs: Log probabilities.
+        mask: Mask.
```
Comment (terrykong) [inline nemo_rl/algorithms/utils.py:198]: docstring is missing logprobs_name

File: nemo_rl/models/policy/workers/dtensor_policy_worker.py [hunk 6]
```diff
@@ -791,7 +816,7 @@ def train(
                                 global_valid_toks,
                             )
                         else:
-                            loss_input = prepare_loss_input(logits, mb, loss_fn)
+                            loss_input, mb = prepare_loss_input(logits, mb, loss_fn)
```
Comment (terrykong) [inline nemo_rl/models/policy/workers/dtensor_policy_worker.py]: do we also need to do the prepare_loss_input_wrapped thing here to pass sampling params?

File: nemo_rl/models/automodel/setup.py [hunk 7]
```diff
@@ -72,8 +73,17 @@ def validate_and_prepare_config(
...
     if "generation" in config and config["generation"] is not None:
-        is_generation_colocated = config["generation"]["colocated"]["enabled"]
+        generation_cfg = config["generation"]
+        # set generation colocated
+        is_generation_colocated = generation_cfg["colocated"]["enabled"]
+        # set sampling params
+        sampling_params = TrainingSamplingParams(
+            top_k=generation_cfg.get("top_k", None),
+            top_p=generation_cfg.get("top_p", 1.0),
+            temperature=generation_cfg.get("temperature", 1.0),
```
Comment (terrykong) [inline nemo_rl/models/automodel/setup.py]: same default comment

File: nemo_rl/models/generation/vllm/vllm_generation.py [hunk 8]
```diff
@@ -42,12 +42,6 @@
     compute_spec_decode_metrics,
 )
 
-# Global thresholds for top_k and top_p validation.
-# While top-k/p are not supported, these values allow for token filtering while the logprobs should be compatible.
-# See https://github.com/NVIDIA-NeMo/RL/issues/69 and https://github.com/NVIDIA-NeMo/RL/issues/237 for more details.
-TOP_K_THRESHOLD = 8000  # Allow top_k >= 8000 (effectively no filtering)
-TOP_P_THRESHOLD = 0.99  # Allow top_p >= 0.99 (close to 1.0)
```
Comment (terrykong) [inline nemo_rl/models/generation/vllm/vllm_generation.py:49]: i feel like we should document https://github.com/NVIDIA-NeMo/RL/issues/237 somewhere. maybe in the docs. by removing this we'll now see errors on some qwen models that had this issue. maybe printing a warning here but also adding it to the top-p/top-k docs as a tip

File: nemo_rl/models/megatron/data.py [hunk 9]
```diff
@@ -551,11 +554,15 @@ def _get_pack_sequence_parameters_for_megatron(
...
+    minimum_pad_factor = 1
     if cp_size > 1:
-        pad_individual_seqs_to_multiple_of *= cp_size * 2
+        minimum_pad_factor *= cp_size * 2
     if tp_size > 1 and sp:
-        pad_individual_seqs_to_multiple_of *= tp_size
+        minimum_pad_factor *= tp_size
+    assert pad_individual_seqs_to_multiple_of % minimum_pad_factor == 0, (
+        f"make_sequence_length_divisible_by ({pad_individual_seqs_to_multiple_of}) is not a multiple of minimum_pad_factor ({minimum_pad_factor}). "
+        "Please set policy.make_sequence_length_divisible_by to a multiple of the minimum_pad_factor."
```
Comment (terrykong) [inline nemo_rl/models/megatron/data.py]: users won't know what this minimum_pad_factor since it's calculation is local. do you think we could provide a more informative error by giving them how that's calculated so they know what to change

File: nemo_rl/models/megatron/train.py [hunk 10]
```diff
@@ -56,7 +61,6 @@
 def model_forward(
     model: GPTModel,
     data_dict: BatchedDataDict[Any],
-    cfg: PolicyConfig,
```
Comment (terrykong) [inline nemo_rl/models/megatron/train.py:59]: nice cleanup!

File: nemo_rl/models/megatron/setup.py [hunk 11]
```diff
@@ -194,12 +195,20 @@ def validate_and_set_config(
     hf_model_name,
     pretrained_path,
     weights_path,
-    tokenizer,
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py:197]: nice cleanup!

File: nemo_rl/algorithms/logits_sampling_utils.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    """
+
+    top_k: int | None = None
+    top_p: float = 1.0
+    temperature: float = 1.0
+
+
+def _need_top_k_filtering(top_k: int | None) -> bool:
+    """Check if top-k filtering is needed."""
+    return top_k is not None and top_k != -1
```
Comment (terrykong) [inline nemo_rl/algorithms/logits_sampling_utils.py]: will this be an issue that mcore inf uses 0?
https://github.com/NVIDIA-NeMo/RL/blob/c19f6d84fe21828b20e413da1827562b65a0b899/nemo_rl/models/policy/workers/megatron_policy_worker.py#L758-L771

File: nemo_rl/models/megatron/setup.py (hunk 2 above)
Comment (yuki-97) [reply nemo_rl/models/megatron/setup.py]: https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/d879b315ef55b43e4623dcbbdebbd6b6c90ad366

File: nemo_rl/algorithms/logits_sampling_utils.py (hunk 12 above)
Comment (yuki-97) [reply nemo_rl/algorithms/logits_sampling_utils.py]: good catch! updated to `top_k > 0`. https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/d879b315ef55b43e4623dcbbdebbd6b6c90ad366

File: nemo_rl/models/policy/workers/dtensor_policy_worker.py (hunk 4 above)
Comment (yuki-97) [reply nemo_rl/models/policy/workers/dtensor_policy_worker.py]: good point, I also think it's not needed. removed in https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/be74105c376c297e6cf62928dc0d928ec9a04bf0

File: nemo_rl/algorithms/utils.py (hunk 5 above)
Comment (yuki-97) [reply nemo_rl/algorithms/utils.py:198]: https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/be74105c376c297e6cf62928dc0d928ec9a04bf0

File: nemo_rl/models/policy/workers/dtensor_policy_worker.py (hunk 6 above)
Comment (yuki-97) [reply nemo_rl/models/policy/workers/dtensor_policy_worker.py]: thanks, good catch! https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/be74105c376c297e6cf62928dc0d928ec9a04bf0

File: nemo_rl/models/megatron/data.py (hunk 9 above)
Comment (yuki-97) [reply nemo_rl/models/megatron/data.py]: https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/be74105c376c297e6cf62928dc0d928ec9a04bf0

File: nemo_rl/algorithms/loss/loss_functions.py (hunk 1 above)
Comment (yuki-97) [reply nemo_rl/algorithms/loss/loss_functions.py:290]: ~~good catch! I also think here should use `curr_logprobs`, updated in https://github.com/NVIDIA-NeMo/RL/pull/2053/changes/be74105c376c297e6cf62928dc0d928ec9a04bf0.~~

feels should use `curr_logprobs_unfiltered` as the comment said it's reweighting samples from π_gen_filtered to π_curr_unfiltered, so I revert in https://github.com/NVIDIA-NeMo/RL/pull/2053/commits/289e1cc675451ca174c4b0c3d4697ed6d861068a.

@zhandaz can you confirm this?

File: nemo_rl/models/generation/vllm/vllm_generation.py (hunk 8 above)
Comment (yuki-97) [reply nemo_rl/models/generation/vllm/vllm_generation.py:49]: added to doc in https://github.com/NVIDIA-NeMo/RL/pull/2053/commits/289e1cc675451ca174c4b0c3d4697ed6d861068a. but I didn't find a good place to add a warning so I didn't add.

File: nemo_rl/algorithms/loss/loss_functions.py (hunk 1 above)
Comment (zhandaz) [reply nemo_rl/algorithms/loss/loss_functions.py:290]: @yuki-97 I think the current implementation is correct. Here since the `kl_importance_weights` is for the kl loss between the current policy and reference model, and we are using the unfiltered logprobs, the kl_importance should also respect that.
https://github.com/NVIDIA-NeMo/RL/blob/a0474b3953ada24a6c6a0681d84eb9dd4f445417/nemo_rl/algorithms/loss/loss_functions.py#L298-L309

**Review by terrykong — APPROVED** [review]

File: nemo_rl/algorithms/loss/loss_functions.py (hunk 1 above)
Comment (terrykong) [reply nemo_rl/algorithms/loss/loss_functions.py:290]: we discussed offline. makes sense. the KL uses unfiltered, so the importance sampling should as well
